Test sticky scheduler.

KonsKo opened the pull request:
Based on [issue ](https://github.com/tempesta-tech/tempesta-test/issues/37)


I also added response parser for curl-like responses. If something same is already exists, please, let me know I am gonna change logic.

const-t commented on http2_general/test_h2_sticky_scheduler.py:
```diff
@@ -0,0 +1,164 @@
+            cache 0;
+            cache_fulfill * *;
+            block_action attack reply;
+            http_chain {
+                host == "bad.com"	-> block;
+                host == "good.com" -> default;
+            }
```
Even, if we will have `Host` header with `good.com` value, request doesn't hit `good.com` vhost, because it routes to `default`. Also, please rename `default` vhost to proper use of `tls_match_any_server_name`

const-t commented on tests_disabled.json:
```diff
@@ -161,10 +161,6 @@
             "name" : "selftests.test_framework.ExampleTest.test_deproxy_srvclient",
             "reason" : "Disabled by issue #73"
         },
-        {
-            "name" : "http2_general",
-            "reason" : "Disabled by issue #261"
-        },
```
We can't to enable whole `http2_general` suite, #261 has not been fixed yet.

KonsKo replied:
oh, sorry, my bad. I uncommented them to work with my test. I returned it back. For my test I created new directory.

KonsKo commented:
@const-t , I reworked whole test - please, check.

KonsKo commented on http2_general/test_h2_sticky_scheduler.py:
```diff
@@ -0,0 +1,164 @@
+
+        curl_init.start()
+        self.wait_while_busy(curl_init)
+        resp_init = parse_response(
```
I tried to use all classes from deproxy (Response, H2Response, Request) - those did not work.

nickzaev commented on helpers/response_parser.py:
```diff
@@ -0,0 +1,69 @@
+"""
+Module to parse curl-like response.
```
I believe it would be better to have this doc-string directly within the `Response` class, this way IDE's have a better chance to provide more related autogenerated docs. Also, the name `Response` doesn't point out the actual purpose of this class and leads to confusion and collisions with, for example, `helpers.deproxy.Response`, so it perhaps should be renamed to something like `helpers.curl.CurlResponse`.

KonsKo replied:
renamed to CurlResponse

nickzaev commented on helpers/response_parser.py:
```diff
@@ -0,0 +1,69 @@
+
+    return Response(
+        http_version=first_line[0],
+        status=first_line[1],
```
Are we 100% sure that in any case, even in the worst one, there always be at least 2 items in `first_line` list?

KonsKo replied:
corrected

nickzaev commented on t_http2/test_h2_sticky_scheduler.py:
```diff
@@ -0,0 +1,164 @@
+"""Test module for http2 and Sticky Cookie Scheduler."""
```
Why do we need a separate directory for this only test? Is there a reason this can't be placed in, say, `./sessions/test_sessions.py`?

KonsKo replied:
I had plans to place tests in directories with prefix, but for now I have no time because main project.
I moved file to dir `/sessions/`.

nickzaev commented on t_http2/test_h2_sticky_scheduler.py:
```diff
@@ -0,0 +1,164 @@
+            tls_certificate ${tempesta_workdir}/tempesta.crt;
+            tls_certificate_key ${tempesta_workdir}/tempesta.key;
+            cache 0;
+            cache_fulfill * *;
```
There's no cache enabled.

KonsKo replied:
corrected

nickzaev commented on t_http2/test_h2_sticky_scheduler.py:
```diff
@@ -0,0 +1,164 @@
+        resp_init = parse_response(resp_init, encoding='latin-1')
+        set_cookie = resp_init.headers.get('set-cookie')
+        self.assertEqual(
+            int(resp_init.status),
```
`status` has already been converted to `int` in `helpers.response_parser.Response.__init__`.

KonsKo replied:
I removed type casting from class, but I see no problem in double type casting

nickzaev commented on t_http2/test_h2_sticky_scheduler.py:
```diff
@@ -0,0 +1,164 @@
+        curl_init.start()
+        self.wait_while_busy(curl_init)
+        resp_init = curl_init.resq.get(True, 1)[0]
+        resp_init = parse_response(resp_init, encoding='latin-1')
```
Why is it `latin-1`? There's no `Accept-Encoding` header provided in the request specifying that coding, also there's no `charset` directive in `nginx` configuration.

KonsKo replied:
There was symbol(s) that did not work with `utf-8` and worked with such encoding. Why - I do not know.

nickzaev commented on t_http2/test_h2_sticky_scheduler.py:
```diff
@@ -0,0 +1,164 @@
+        """,
+    }
+
+    backends = [
```
Wouldn't it be sufficient to just use `deproxy` backend here?

KonsKo replied:
I have no answer. Perhaps, deproxy did not work with h2 at the moment when I started this task.

nickzaev reviewed (changes requested):
At the moment, this test fails with:
```
ERROR: test_h2_cookie_scheduler (t_http2.test_h2_sticky_scheduler.H2StickySchedulerTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/tempesta-test/t_http2/test_h2_sticky_scheduler.py", line 115, in test_h2_cookie_scheduler
    resp_init = parse_response(resp_init, encoding='latin-1')
  File "/root/tempesta-test/helpers/response_parser.py", line 66, in parse_response
    http_version=first_line[0],
IndexError: list index out of range

----------------------------------------------------------------------
Ran 1 test in 1.470s

FAILED (errors=1)
```

KonsKo commented on helpers/response_parser.py:
```diff
@@ -0,0 +1,69 @@
+from typing import Union, Optional
+
+
+class Response(object):
+
+    def __init__(
+        self,
+        http_version: str = '',
+        status: Optional[Union[str, int]] = None,
+        headers: Optional[dict] = None,
+        error: Optional[str] = None,
+    ):
+        super().__init__()
+        self.http_version = http_version
+        self.status = int(status)
+        self.headers = headers if headers else {}
+        self.error = error if error else ''
+
+    def __repr__(self):
+        return '{0} {1};\nHeaders: {2}\nErrors: {3}'.format(
+            self.http_version,
+            self.status,
+            self.headers,
+            self.error,
+        )
```
About explicit base class: this has nothing to do with python2. It is only about consistency.
I removed 'super' - It could left from first version of PR, I already did not remember purpose of it and I fully agree It should be removed from here.
About dataclass: it is only simplified work. In the future, we will be able to want to change something(add validation or whatever). Work with class without any layer. I do not want to use dataclass here.

KonsKo commented on helpers/response_parser.py:
```diff
@@ -0,0 +1,69 @@
+    string_to_parse = string_to_parse.split('\r\n')
+
+    # we expect first line such 'HTTP/1.1 200 OK'
+    try:
```
agree, removed

KonsKo commented on helpers/response_parser.py:
```diff
@@ -0,0 +1,69 @@
+    # we expect first line such 'HTTP/1.1 200 OK'
+    try:
+        first_line = string_to_parse[0].split()
+    except Exception as exc:
```
already removed because comment above

KonsKo commented:
@nickzaev, I made some corrections. Sorry, that too late. Please, take a look when you have time.

b3b reviewed (changes requested):
Test fails:
```
./run_tests.py sessions.test_h2_sticky_scheduler
======================================================================
ERROR: test_h2_cookie_scheduler (sessions.test_h2_sticky_scheduler.H2StickySchedulerTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/tempesta-test2/sessions/test_h2_sticky_scheduler.py", line 116, in test_h2_cookie_scheduler
    set_cookie = resp_init.headers.get('set-cookie')
AttributeError: 'NoneType' object has no attribute 'headers'

======================================================================
ERROR: test_h2_cookie_scheduler (sessions.test_h2_sticky_scheduler.H2StickySchedulerTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/tempesta-test2/framework/tester.py", line 358, in tearDown
    if self.oops._warn_count(err) > 0:
  File "/root/tempesta-test2/helpers/dmesg.py", line 53, in _warn_count
    return len(self._warn_match(warn_str))
  File "/root/tempesta-test2/helpers/dmesg.py", line 56, in _warn_match
    return re.findall(warn_str, self.log.decode())
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x84 in position 273: invalid start byte

----------------------------------------------------------------------
```

Reason for the failure is corrupted location header in response.
I was trying to change protocol to HTTPS (test should pass), but still test fails because of some parsing errors.

Suggestions:
- HTTP2 test should be disabled
- Let's change test to use CurlClient (https://github.com/tempesta-tech/tempesta-test/pull/332)? It has cookies support
- Let's also add HTTP/1 and HTTPS tests? Sticky cookies works differently with different protocols.

b3b commented:
This could be done in separate PR, #37 requirement is not satisfied: `The scheduler must also be tested for learn sticky cookie mode.`

b3b commented:
And about test logic,
step 5: "client requests URI http://bad.com/ with sticky cookie set". 
Shouldn't it be  "client requests URI http://bad.com/ **without** sticky cookie set", to check that request not blocked but redirected?

KonsKo commented:
> Suggestions:
> 
> * HTTP2 test should be disabled
> * Let's change test to use CurlClient (#332)? It has cookies support
> * Let's also add HTTP/1 and HTTPS tests? Sticky cookies works differently with different protocols.

I changed test as you suggested (I copied-pasted your code) and I removed unrelated earlier created module.
About adding of support HTTP and HTTPS  - I think it is not in scope of #37 
I changed formatting.

nickzaev reviewed (approved):
LGTM
